nomt/src/rw_pass_cell.rs

```diff
@@ -90,61 +90,213 @@ impl RwPassDomain {
+        sync_flag.store(true, Ordering::Release);
+
+        WritePassEnvelope {
+            inner: self,
+            sync_flag,
+        }
+    }
+}
+
+impl<R: Region + Clone> WritePass<R> {
+    /// Split this write pass into two parts encompassing non-overlapping sub-regions.
+    ///
+    /// The result will be a `(left, right)` tuple corresponding to the input argument regions.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the regions overlap with each other or are not encompassed by the region of this
+    /// pass.
+    pub fn split(mut self, left: R, right: R) -> (Self, Self) {
+        assert!(self.region().encompasses(&left));
+        assert!(self.region().encompasses(&right));
+        assert!(
+            left.excludes_unique(&right),
+            "left and right regions overlap"
+        );
+
+        let new_parent = Arc::new(ParentWritePass {
+            parent: self.parent.take(),
+            region: self.region().clone(),
+            remaining_children: AtomicUsize::new(2),
+        });
+
+        let left_pass = WritePass {
+            parent: Some(new_parent.clone()),
+            read_pass: ReadPass {
+                domain: self.read_pass.domain.clone(),
+                region: left,
+                _guard: self.read_pass._guard.clone(),
+            },
+        };
+
+        let right_pass = WritePass {
+            parent: Some(new_parent),
+            read_pass: ReadPass {
+                domain: self.read_pass.domain.clone(),
+                region: right,
+                _guard: self.read_pass._guard.clone(),
+            },
+        };
+
+        (left_pass, right_pass)
+    }
+
+    /// Consume this regioned write-pass, possibly yielding the parent region back.
+    ///
+    /// If all other split descendents of the parent write pass have been dropped or consumed,
+    /// this will return a region equivalent to the parent's.
+    pub fn consume(self) -> Option<Self> {
+        let Some(ref parent) = self.parent else {
+            return None;
+        };
+
+        // SAFETY: release our writes to other threads if _not_ the last sibling.
+        //         acquire writes from other threads if the last sibling.
+        //         the value of '1' must have been written by another thread with `Release`.
+        if parent.remaining_children.fetch_sub(1, Ordering::AcqRel) == 1 {
+            Some(WritePass {
+                parent: parent.parent.clone(),
+                read_pass: ReadPass {
+                    domain: self.read_pass.domain.clone(),
+                    region: parent.region.clone(),
+                    _guard: self.read_pass._guard.clone(),
+                },
+            })
+        } else {
+            None
+        }
+    }
 }
 
+impl WritePass<UniversalRegion> {
+    /// Supply a region type.
+    pub fn with_region<R>(self, region: R) -> WritePass<R> {
+        // sanity: UniversalRegion can't be split, so this should never be Some.
```

Comment:
The implementation of `RegionContains` returns true for both `contains` and `contains_exclusive`. Why `split` should fail if called on `WritePass<UniversalRegion>` ? The two asserts seems to be satisfied at the beginning of `split`